<a href="https://colab.research.google.com/github/1nvl/EDA-Superstore/blob/main/Tutu.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [60]:
#Импортируем библиотеку Пандас для работы с данными, прочитаем csv файл и выведем наш датасет:
import pandas as pd
df = pd.read_csv('ha_data.csv',';')
df

<ipython-input-60-983cb68c8b9d>:3: FutureWarning: In a future version of pandas all arguments of read_csv except for the argument 'filepath_or_buffer' will be keyword-only.
  df = pd.read_csv('ha_data.csv',';')


order_id     page_id           product site_version  \
0             NaN         3.0           company       mobile   
1             NaN   3699687.0           company       mobile   
2             NaN        14.0          sneakers       mobile   
3             NaN  10362176.0           company       mobile   
4             NaN        14.0          sneakers       mobile   
...           ...         ...               ...          ...   
8471221       NaN  24897184.0       accessories      desktop   
8471222       NaN  24897190.0           clothes       mobile   
8471223       NaN  24897203.0  sports_nutrition       mobile   
8471224       NaN  24897205.0          sneakers       mobile   
8471225       NaN  24897215.0       accessories       mobile   

                        time         title          user  
0        2017-02-09 20:24:04   banner_show        user_0  
1        2017-02-07 10:03:07   banner_show        user_0  
2        2017-01-29 13:02:23   banner_show        user_1  
3        2017-04-12 15:39:19   banner_show        user_1  
4        2017-01-29 13:04:42  banner_click        user_1  
...                      ...           ...           ...  
8471221  2017-05-23 14:07:00   banner_show  user_4254616  
8471222  2017-05-28 08:10:20   banner_show  user_4254617  
8471223  2017-05-20 09:20:50   banner_show  user_4254618  
8471224  2017-05-28 19:25:42   banner_show  user_4254619  
8471225  2017-05-31 09:18:24   banner_show  user_4254620  

[8471226 rows x 7 columns]

In [2]:
#Кол-во показов, кликов и покупок для каждого нашего баннера за все время:
banner_counts = df.groupby('product')['title'].value_counts().unstack(fill_value=0)
print(banner_counts)

title             banner_click  banner_show   order
product                                            
accessories             136254      1440954   44551
clothes                 220595      1453128  112715
company                 144821      1580238       0
sneakers                182597      1453029   67719
sports_nutrition        144918      1465970   23737


In [3]:
#Показатель кликабельности(CTR) для каждого продукта за все время:
banner_ctr = banner_counts['banner_click']/banner_counts['banner_show']*100
print(banner_ctr)

product
accessories          9.455819
clothes             15.180700
company              9.164506
sneakers            12.566645
sports_nutrition     9.885468
dtype: float64


In [12]:
#Конверсия в покупку (CR) для каждого продукта за все время:
banner_cr = banner_counts['order']/banner_counts['banner_show']*100
print(banner_cr)

product
accessories         3.091771
clothes             7.756715
company             0.000000
sneakers            4.660540
sports_nutrition    1.619201
dtype: float64


In [11]:
#Какой процент кликов приводит к покупке для каждого продукта за все время:
banner_сlick_to_order = banner_counts['order']/banner_counts['banner_click']*100
print(banner_сlick_to_order)

product
accessories         32.697022
clothes             51.095900
company              0.000000
sneakers            37.086590
sports_nutrition    16.379608
dtype: float64


In [63]:
#Кол-во показов, кликов и покупок всех продуктов на каждой версии сайта за все время:
banner_site_version = df.groupby('site_version')['title'].value_counts().unstack(fill_value=0)
print(banner_site_version)

title         banner_click  banner_show   order
site_version                                   
desktop             115065      2134640  133181
mobile              714120      5258679  115541


In [14]:
#Показатель кликабельности (CTR) для каждой версии сайта:
site_version_ctr = banner_site_version['banner_click']/banner_site_version['banner_show']*100
print(site_version_ctr)


site_version
desktop     5.390370
mobile     13.579836
dtype: float64


In [15]:
#Конверсия в покупку (CR) для каждой версии сайта:
site_version_cr = banner_site_version['order']/banner_site_version['banner_show']*100
print(site_version_cr)


site_version
desktop    6.239038
mobile     2.197149
dtype: float64


In [ ]:
#Сортировка по time по убыванию:
df.sort_values(by='time',ascending=False)

In [17]:
#Преобразуем столбец time в формат даты и времени, выведем все записи последнего месяца (май - последний месяц):
df['time'] = pd.to_datetime(df['time'])
df['month'] = df['time'].dt.month
df_may = df[df['month']==5]
df_may.sort_values(by='time',ascending=False)

order_id     page_id           product site_version  \
7880258  354948.0         NaN           clothes      desktop   
7808442       NaN  22326691.0           clothes       mobile   
8363442       NaN  24446787.0  sports_nutrition       mobile   
8165105       NaN  23658876.0           clothes      desktop   
8151182       NaN  23601866.0           company       mobile   
...           ...         ...               ...          ...   
6298645       NaN  16922876.0           company       mobile   
8022916       NaN  23114628.0           company      desktop   
7164663  305073.0         NaN       accessories      desktop   
5245930       NaN  13783299.0           company       mobile   
5624435       NaN  22816201.0           company      desktop   

                       time         title          user  month  
7880258 2017-05-31 23:59:58         order  user_3884986      5  
7808442 2017-05-31 23:59:53   banner_show  user_3843064      5  
8363442 2017-05-31 23:59:51  banner_click  user_4182938      5  
8165105 2017-05-31 23:59:40   banner_show  user_4056326      5  
8151182 2017-05-31 23:59:37   banner_show  user_4047676      5  
...                     ...           ...           ...    ...  
6298645 2017-05-01 00:00:15   banner_show  user_2994842      5  
8022916 2017-05-01 00:00:10   banner_show  user_3969556      5  
7164663 2017-05-01 00:00:08         order  user_3474533      5  
5245930 2017-05-01 00:00:07   banner_show  user_2457220      5  
5624435 2017-05-01 00:00:02   banner_show  user_2643547      5  

[2130108 rows x 8 columns]

In [19]:
#Кол-во показов, кликов и покупок для каждого нашего баннера в мае:
banner_counts_may = df_may.groupby('product')['title'].value_counts().unstack(fill_value=0)
print(banner_counts_may)

title             banner_click  banner_show  order
product                                           
accessories              34787       361345   9381
clothes                  56870       372245  23481
company                  34809       378538      0
sneakers                 47088       370090  14498
sports_nutrition         38039       383933   5004


In [20]:
#Конверсия в покупку (CR) по всем категориям в мае:
banner_may_cr = banner_counts_may['order']/banner_counts_may['banner_show']*100
print(banner_may_cr)

product
accessories         2.596134
clothes             6.307942
company             0.000000
sneakers            3.917425
sports_nutrition    1.303352
dtype: float64


In [21]:
#Аналогично найдем конверсии в покупку (CR) по всем категориям для остальных месяцев:
#Январь:
df_jan = df[df['month']==1]
banner_counts_jan = df_jan.groupby('product')['title'].value_counts().unstack(fill_value=0)
banner_jay_cr = banner_counts_jan['order']/banner_counts_jan['banner_show']*100
#Ферваль:
df_feb = df[df['month']==2]
banner_counts_feb = df_feb.groupby('product')['title'].value_counts().unstack(fill_value=0)
banner_feb_cr = banner_counts_feb['order']/banner_counts_feb['banner_show']*100
#Март:
df_mar = df[df['month']==3]
banner_counts_mar = df_mar.groupby('product')['title'].value_counts().unstack(fill_value=0)
banner_mar_cr = banner_counts_mar['order']/banner_counts_mar['banner_show']*100
#Апрель:
df_apr = df[df['month']==4]
banner_counts_apr = df_apr.groupby('product')['title'].value_counts().unstack(fill_value=0)
banner_apr_cr = banner_counts_apr['order']/banner_counts_apr['banner_show']*100
#Выведем все конверсии в покупки (CR) по этим месяцам:
print('Январь:','\n',banner_jay_cr,'\n')
print('Февраль:','\n',banner_feb_cr,'\n')
print('Март:','\n',banner_mar_cr,'\n')
print('Апрель:','\n',banner_apr_cr,'\n')

Январь: 
 product
accessories         2.444307
clothes             6.387628
company             0.000000
sneakers            3.777081
sports_nutrition    1.376356
dtype: float64 

Февраль: 
 product
accessories          3.793953
clothes             10.054882
company              0.000000
sneakers             5.855666
sports_nutrition     2.012863
dtype: float64 

Март: 
 product
accessories         3.422065
clothes             8.568465
company             0.000000
sneakers            5.133593
sports_nutrition    1.835512
dtype: float64 

Апрель: 
 product
accessories         3.403558
clothes             8.316737
company             0.000000
sneakers            5.012631
sports_nutrition    1.739670
dtype: float64 



In [23]:
#Кол-во показов, кликов и покупок для каждого нашего баннера в феврале:
print(banner_counts_feb)

title             banner_click  banner_show  order
product                                           
accessories              19043       201294   7637
clothes                  30724       200430  20153
company                  26975       298805      0
sneakers                 25537       201241  11784
sports_nutrition         20034       201504   4056


In [24]:
#Кол-во уникальных пользоваталей, видивших какой-либо из баннеров:
users_count = df['user'].nunique()
print(users_count)

4254621


In [36]:
#Кол-во пользователей, совершивших хотя бы 1 покупку за все время, после клика на баннер:
filt_df = df[df['title'] == 'order']
users_count_order = filt_df['user'].nunique()
print(users_count_order)

237866


In [49]:
#Кол-во пользователей, совершавших более 1 покупки за все время, после клика на баннер:
filt_df1 = filt_df.groupby('user').filter(lambda x: len(x)>1)
users_count_order1 = filt_df1['user'].nunique()
print(users_count_order1)

10001


In [52]:
#Кол-во пользователей, совершавших более 5 покупок за все время, после клика на баннер:
filt_df5 = filt_df.groupby('user').filter(lambda x: len(x)>5)
users_count_order5 = filt_df5['user'].nunique()
print(users_count_order5)

1


In [53]:
#Пользователь с наибольшим числом покупок после кликов на баннер:
user_order_counts = df[df['title'] == 'order']['user'].value_counts()
max_user = user_order_counts.idxmax()
print(max_user)

user_674959


In [54]:
#В каком часу совершалось наибольшее число покупок после кликов на баннер:
purchase_hour_counts = df[df['title'] == 'order']['time'].dt.hour.value_counts()
max_purchase_hour = purchase_hour_counts.idxmax()
print(max_purchase_hour)

15


In [26]:
df.sort_values(by='order_id',ascending=False)

order_id     page_id           product site_version  \
7018316  855750.0         NaN          sneakers       mobile   
5991010  855748.0         NaN          sneakers       mobile   
598355   855741.0         NaN          sneakers       mobile   
7108136  855737.0         NaN           clothes       mobile   
3814318  855735.0         NaN           clothes       mobile   
...           ...         ...               ...          ...   
8471221       NaN  24897184.0       accessories      desktop   
8471222       NaN  24897190.0           clothes       mobile   
8471223       NaN  24897203.0  sports_nutrition       mobile   
8471224       NaN  24897205.0          sneakers       mobile   
8471225       NaN  24897215.0       accessories       mobile   

                       time        title          user  month  
7018316 2017-03-31 23:07:53        order  user_3391741      3  
5991010 2017-04-11 18:49:09        order  user_2830713      4  
598355  2017-02-07 17:48:59        order   user_234843      2  
7108136 2017-04-27 03:16:09        order  user_3442201      4  
3814318 2017-04-23 23:52:48        order  user_1775727      4  
...                     ...          ...           ...    ...  
8471221 2017-05-23 14:07:00  banner_show  user_4254616      5  
8471222 2017-05-28 08:10:20  banner_show  user_4254617      5  
8471223 2017-05-20 09:20:50  banner_show  user_4254618      5  
8471224 2017-05-28 19:25:42  banner_show  user_4254619      5  
8471225 2017-05-31 09:18:24  banner_show  user_4254620      5  

[8471226 rows x 8 columns]

In [31]:
#Общее количество покупок в нашем онлайн магазине (речь идет вообще о всех покупках в нашем онлайн магазине, т.е. не только те покупки, которые совершались после клика на баннер):
all_orders = int(df['order_id'].max())
print(all_orders)

855750


In [32]:
#Количество покупок после клика на баннер:
banner_orders = df['title'].value_counts()['order']
print(banner_orders)

248722


In [39]:
#Количество покупок, совершенных, не взаимодействуя с баннерами:
print(all_orders-banner_orders)

607028


In [62]:
#Конверсия в покупку (CR) за все 5 месяцев для тех покупок, которые совершались после клика на баннер:
print(banner_orders/users_count*100)

5.845926111867543


In [66]:
df.sort_values(by='page_id')
#Найдем общее количество пользователей, которые посещали наш онлайн магазин(т.е. не только те пользователи, которые видели баннер, а вообще все):
all_users = int(df['page_id'].max())
print(all_users)

24897219


In [65]:
#Найдем количество пользователей совершивших покупку, не взаимодействуя с баннером:
print(all_users-users_count)

20642598


In [64]:
#Конверсия в покупку (CR) за все 5 месяцев для тех покупок, которые совершались без взаимодействия с баннерами:
print((all_orders-banner_orders)/(all_users-users_count)*100)

2.9406569851333635


Вывод: продавать место для рекламы другого сервиса не нужно. Конверсия в покупку за все 5 месяцев для тех покупок, которые совершались без взаимодействия с баннерами ниже на 2.9% по сравнению с конверсией в покупку тех покупок в тот же временной период, которые совершались после кликов на баннеры.
Однако стоит изменить следующее:
1) Баннер company показывается чаще остальных наших баннеров (1580238 раз за 5 месяцев), однако какой в этом смысл, если он не продающий и мы не получаем с него orderов. Его количеcтво показов я бы уменьшил, а вместо этого увеличил показы баннера clothes, которые демонстрирует наилучшую конверсию в покупку каждый месяц при том же количестве показов, что и у остальных наших баннеров. Либо вместо баннера company вообще показывать баннер какого-либо нового продукта.
2) Далее стоит обратить внимание на версии сайтов. В мобильной версии мы показываем баннеры в 2.5 раза чаще, но при этом конверсия в покупку почти в 3 раза выше у десктопной версии сайта. Из этого следует вывод, что количество показов баннеров в десктопной версии нужно увеличить.
3) Конверсия в покупку по месяцам для тех покупок, которые совершались после взаимодействия с баннерами тоже растет. Например, виден явный прогресс в феврале и марте по сравнению с январем, конверсия по всем продуктам стала выше (от 0.5% до 4% повышения).
Интересные факты из нашего датафрейма:
1) Пользователь с наибольшим числом покупок после кликов на баннер - это user_674959.
2) C 15:00 до 16:00 совершалось наибольшее число покупок после кликов на баннер.
3) Количество пользователей, совершавших более 1 покупки за все время, путем взаимодействия с баннерами - это 10001 человек, однако более 5 покупок, взаимодействуя с баннерами совершил только 1 пользователь.
